### PR TITLE
Add support for NI with multiple ports

### PR DESCRIPTION
**Pull request by milan-zededa**
This commit implements support for Local NI with multiple ports attached.
API changes with detailed description can be found here:
https://github.com/lf-edge/eve-api/pull/53
This is a continuation to PRs:
https://github.com/lf-edge/eve/pull/4098
https://github.com/lf-edge/eve/pull/4113
(next PR will bring documentation updates)

In summary, network instance can be now configured with "shared" port
label, potentially matching multiple device ports. The NI routing table
will contain routes from all the selected ports.

Shared labels can be also used to restrict port-forwarding to a subset
of NI ports and to create multipath static routes (routes with multiple
possible next-hops). For every multipath route, zedrouter will use
recently added portprober to select the best port at a given time (based
on the connectivity status, cost, etc.) and also to failover to another
port when the currently used port looses connectivity.

Eden-SDN example for NI with multiple ports: https://github.com/lf-edge/eden/pull/1008

### CODE REVIEW

File: pkg/pillar/cmd/zedrouter/zedrouter.go [hunk 1]
```diff
@@ -215,8 +222,19 @@
...
+		return fmt.Errorf("failed to read the controller hostname from file %s: %v",
+			types.ServerFileName, err)
+	}
+	z.controllerHostname = string(content)
+	z.controllerHostname = strings.TrimSpace(z.controllerHostname)
+	z.controllerPort = 443
+	if host, port, err := net.SplitHostPort(z.controllerHostname); err == nil {
+		z.controllerHostname = host
+		if portNum, err := strconv.Atoi(port); err == nil {
+			z.controllerPort = uint16(portNum)
```
[the remark this answers is not in the document]
Comment (milan-zededa) [reply pkg/pillar/cmd/zedrouter/zedrouter.go:236]: Fixed

File: pkg/pillar/cmd/diag/diag.go [hunk 2]
```diff
@@ -880,7 +880,7 @@ func printOutput(ctx *diagContext, caller string) {
 		// fields are set and Dhcp type; proxy info order
 		ifname := port.IfName
 		isMgmt := types.IsMgmtPort(*ctx.DeviceNetworkStatus, ifname)
-		priority := types.GetPortCost(*ctx.DeviceNetworkStatus,
+		cost := types.GetPortCost(*ctx.DeviceNetworkStatus,
```
Comment (eriknordmark) [inline pkg/pillar/cmd/diag/diag.go:883]: I didn't know we were using "priority" in some cases. Thanks for making it consistently use "cost" and not mixing the two.

File: pkg/pillar/cmd/zedrouter/networkinstance.go [hunk 3]
```diff
@@ -81,124 +90,259 @@ func (z *zedrouter) getNIBridgeConfig(
...
-		return fmt.Errorf("missing uplink interface '%s'", ifName)
-	}
-	if status.IsUsingUplinkBridge() {
-		_, ifMAC, _ := z.networkMonitor.GetInterfaceAddrs(ifIndex)
-		status.BridgeMac = ifMAC
-	}
-	return nil
+	newNTPServers = generics.FilterDuplicatesFn(newNTPServers, netutils.EqualIPs)
+	changed = changed || !generics.EqualSets(status.Ports, newPortLabels)
+	status.Ports = newPortLabels
```
Comment (naiming-zededa) [inline pkg/pillar/cmd/zedrouter/networkinstance.go:142]: Just a question, when the config for the NI only has one label, the "PortLabel", and here the status will have the array of Ports which is also labels, that means the NI status will have multiple labels?

File: pkg/pillar/cmd/zedrouter/pubsubhandlers.go [hunk 4]
```diff
@@ -122,11 +122,50 @@ func (z *zedrouter) handleDNSImpl(ctxArg interface{}, key string,
...
+		if niConfig.Activate && !niStatus.Activated && niStatus.EligibleForActivate() {
+			z.doActivateNetworkInstance(*niConfig, &niStatus)
+			z.checkAndRecreateAppNetworks(niStatus.UUID)
+		}
+		z.publishNetworkInstanceStatus(&niStatus)
 	}
 
-	if z.uplinkProber != nil {
-		z.uplinkProber.ApplyDNSUpdate(status)
+	if z.portProber != nil {
```
Comment (naiming-zededa) [inline pkg/pillar/cmd/zedrouter/pubsubhandlers.go:167]: wouldn't this z.portProber already exist always if we receive the DNS?

File: pkg/pillar/cmd/zedrouter/zedrouter.go [hunk 5]
```diff
@@ -215,8 +222,19 @@ func (z *zedrouter) init() (err error) {
 	gcp := *types.DefaultConfigItemValueMap()
 	z.appContainerStatsInterval = gcp.GlobalValueInt(types.AppContainerStatsInterval)
 
+	content, err := os.ReadFile(types.ServerFileName)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to read the controller hostname from file %s: %v",
+			types.ServerFileName, err)
+	}
+	z.controllerHostname = string(content)
```
Comment (naiming-zededa) [inline pkg/pillar/cmd/zedrouter/zedrouter.go:230]: For some of us, using local cluster, we do port mapping on the server, and the server file has <hostname:port> syntax. May need to detect if this type of syntax is used in the server file.

File: pkg/pillar/cmd/zedagent/handlenetworkinstance.go [hunk 6]
```diff
@@ -81,9 +81,22 @@ func prepareAndPublishNetworkInstanceInfoMsg(ctx *zedagentContext,
 	if !deleted {
 		info.Displayname = status.DisplayName
 		info.InstType = uint32(status.Type)
-		info.CurrentUplinkIntf = status.SelectedUplinkIntfName
+		info.Ports = status.Ports
+		if len(status.Ports) > 0 {
+			// For backward-compatibility with older controllers.
+			// Just report the first port from the list.
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/handlenetworkinstance.go:87]: When I read the code I didn't realize that it reports all ports using info.Ports and one for backwards compatibility using the old CurrentUplinkIntf.
How about moving the comment above where you set info.Ports and explaining that it is for new controllers and CurrentUplinkIntf for old in one comment?

File: pkg/pillar/cmd/zedagent/parseconfig.go [hunk 7]
```diff
@@ -418,19 +418,34 @@ func parseStaticRoute(route *zconfig.IPRoute, config *types.NetworkInstanceConfi
...
+	customProbe, err := parseConnectivityProbe(route.GetProbe().GetCustomProbe())
+	if err != nil {
+		return fmt.Errorf("invalid connectivity probe config: %v", err)
 	}
 	config.StaticRoutes = append(config.StaticRoutes, types.IPRouteConfig{
-		DstNetwork: dstNetwork,
-		Gateway:    gatewayIP,
+		DstNetwork:      dstNetwork,
+		Gateway:         gatewayIP,
+		OutputPortLabel: route.Port,
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/parseconfig.go:441]: Does it make sense to validate route.Port here (just like we validate aka parse the gateway) before using it? Or have zedrouter return an error should it not find the port?

File: pkg/pillar/cmd/zedrouter/pubsubhandlers.go [hunk 8]
```diff
@@ -146,6 +185,28 @@ func (z *zedrouter) handleDNSDelete(ctxArg interface{}, key string,
 	z.log.Functionf("handleDNSDelete done for %s", key)
 }
 
+func (z *zedrouter) handleWwanMetricsCreate(ctxArg interface{}, key string,
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedrouter/pubsubhandlers.go:188]: How do there Wwan metrics relate to the multiple ports work?

File: pkg/pillar/cmd/zedrouter/zedrouter.go (hunk 5 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/zedrouter/zedrouter.go:230]: @naiming-zededa looks like that is handled below using SplitHostPort.

Comment (naiming-zededa) [reply pkg/pillar/cmd/zedrouter/zedrouter.go:230]: Ok

File: pkg/pillar/cmd/zedrouter/networkinstance.go (hunk 3 above)
Comment (milan-zededa) [reply pkg/pillar/cmd/zedrouter/networkinstance.go]: In the config, the port label can be logical, matching one port, or a shared label, potentially matching multiple ports.
Whereas in the status we report logical ports of attached ports (after the shared label is resolved).
If NI has only one port attached, this list will have only one entry with the logical label of the port.
I have renamed `newPortLabels` to `newPortLogicalLabels` to make it clear that we do not store shared labels in the `status.Ports` list.

File: pkg/pillar/cmd/zedrouter/pubsubhandlers.go (hunk 4 above)
Comment (milan-zededa) [reply pkg/pillar/cmd/zedrouter/pubsubhandlers.go]: Good point, I have removed the unnecessary check.

File: pkg/pillar/cmd/zedagent/handlenetworkinstance.go (hunk 6 above)
Comment (milan-zededa) [reply pkg/pillar/cmd/zedagent/handlenetworkinstance.go:89]: Done

File: pkg/pillar/cmd/zedagent/parseconfig.go (hunk 7 above)
Comment (milan-zededa) [reply pkg/pillar/cmd/zedagent/parseconfig.go:441]: There is a mechanism in zedrouter to mark the NI with an error when a route has port label configured but it doesn't match any port (with the currently applied DPC). Still the NI is activated, just the problematic route(s) is/are skipped.
When DPC is changed and route is suddenly matching at least one port, the error is cleared and the route is added into the routing table.
The reason why we handle this differently to e.g. gateway IP validation, is that port missing can be temporary issue (e.g. due to a DPC fallback) and the error can be later cleared.

File: pkg/pillar/cmd/zedrouter/pubsubhandlers.go (hunk 8 above)
Comment (milan-zededa) [reply pkg/pillar/cmd/zedrouter/pubsubhandlers.go:186]: This was requested by a customer: https://github.com/lf-edge/eve-api/blob/main/proto/config/netinst.proto#L121-L125
With multiple modems, portprober will prefer modem with better network signal.
Zedrouter needs to subscribe to wwan metrics with signal strength info and pass that to the portprober.

**Comment by milan-zededa** [comment]
> FWIW I see "Uplink" and "UplinkProber" in pkg/pillar/docs/zedrouter.md and in pkg/pillar/types/domainmgrtypes.go. Don't know if/when you want to change those.

There was an unused `SwitchUplink` attribute of `VlanInfo` in domainmgrtypes.go - I have removed it.
Documentation update will come in a separate PR. I have it already prepared in my EVE fork.

File: pkg/pillar/cmd/zedagent/parseconfig.go (hunk 7 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/zedagent/parseconfig.go:441]: Ok

File: pkg/pillar/cmd/zedrouter/pubsubhandlers.go (hunk 8 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/zedrouter/pubsubhandlers.go:186]: Ah - I just didn't see it as a top-level description but the "cost etc" is referencing signal strength. Got it.

**Pull request merged**